publisher/publisher_test.go

```diff
@@ -0,0 +1,198 @@
+RBXq2AcD8Ej7AIU15dRnLEkESmJm4xtV2aqmCd0SSBGhJHYLcInUPzWVg1zcB5EQ
+78GOTue8UrZvbcYhOufHG0k5JX5HVoVZ6GSXKqn5kqbcHXT6adVoWT/BxZruZiKQ
+qkryoZoSywt7dDdDhpC2+oAOC+XwX2HJp2mrPaAea1+E4LM9C9iEDtjsn5FfsBz0
+VRbMRdaoayXzOlTRhF3pGU2LLCmrXy/pqpqAGYPxyHr3auRn9fjv77UMEqVFdfOc
+CspkK71IGqM9UwwMtCZBp0fK/Xv9o1d85paXcJ/aH8zg6EK4UkuXDFnLsg1LrIru
++YHeHOeSaXJlcjzwWVY/Exe5HymtqGH8klMhy65bjtapNt76+j2CJgxOdPEiTy/l
+9LH5ujlo5qgemXE3ePwYZ9D3iiJThTf3tWkvdbz2wCPJAy2EHS0FxHMfx5sXsFsa
+OY8B7wwvZTLzU6WWs781TJXx2CE04PneeeArLpVLkiGIWjk=
+-----END CERTIFICATE-----`
+
+const issuerPath = "../test/test-ca.pem"
+
+func logSrv(t *testing.T, stopChan, waitChan chan bool) {
+	// Reset any existing handlers
+	http.DefaultServeMux = http.NewServeMux()
+
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		decoder := json.NewDecoder(r.Body)
+		var jsonReq ctSubmissionRequest
+		err := decoder.Decode(&jsonReq)
+		if err != nil {
+			return
+		}
+		// Submissions should always contain at least one cert
+		if len(jsonReq.Chain) >= 1 {
+			fmt.Fprint(w, `{"signature":"BAMASDBGAiEAknaySJVdB3FqG9bUKHgyu7V9AdEabpTc71BELUp6/iECIQDObrkwlQq6Azfj5XOA5E12G/qy/WuRn97z7qMSXXc82Q=="}`)
+		}
+	})
+
+	server := &http.Server{Addr: "localhost:8080"}
+	conn, err := net.Listen("tcp", server.Addr)
+	if err != nil {
+		waitChan <- true
+		t.Fatalf("Couldn't listen on %s: %s", server.Addr, err)
+	}
+
+	go func() {
+		<-stopChan
+		conn.Close()
+	}()
+
+	waitChan <- true
+	server.Serve(conn)
+}
+
+func TestNewPublisherAuthorityImpl(t *testing.T) {
+	// Allowed
+	ctConf := CTConfig{SubmissionBackoffString: "0s", BundleFilename: issuerPath}
+	_, err := NewPublisherAuthorityImpl(&ctConf)
+	test.AssertNotError(t, err, "Couldn't create new PublisherAuthority")
+
+	ctConf = CTConfig{Logs: []LogDescription{LogDescription{URI: "http://localhost:8080/ct/v1/add-chain"}}, SubmissionBackoffString: "0s", BundleFilename: issuerPath}
+	_, err = NewPublisherAuthorityImpl(&ctConf)
+	test.AssertNotError(t, err, "Couldn't create new PublisherAuthority")
+}
+
+func TestCheckSignature(t *testing.T) {
+	// Based on a submission to the aviator log
+	goodSigBytes, err := base64.StdEncoding.DecodeString("BAMASDBGAiEAknaySJVdB3FqG9bUKHgyu7V9AdEabpTc71BELUp6/iECIQDObrkwlQq6Azfj5XOA5E12G/qy/WuRn97z7qMSXXc82Q==")
+	test.AssertNotError(t, err, "Couldn't decode signature")
+
+	testReceipt := core.SignedCertificateTimestamp{
+		Signature: goodSigBytes,
+	}
+
+	// Good signature
+	err = testReceipt.CheckSignature()
+	test.AssertNotError(t, err, "Valid signature check failed")
+
+	// Invalid signature (too short, trailing garbage)
+	testReceipt.Signature = goodSigBytes[1:]
+	err = testReceipt.CheckSignature()
+	test.AssertError(t, err, "Invalid signature check failed")
+	testReceipt.Signature = append(goodSigBytes, []byte{0, 0, 1}...)
+	err = testReceipt.CheckSignature()
+	test.AssertError(t, err, "Invalid signature check failed")
+}
+
+func TestSubmitToCT(t *testing.T) {
```

Comment:
Needs test cases for:
- network failure, retry with success
- network failure, retry until max out retry limit
- non-retryable HTTP status code
- retryable HTTP status code with a Retry-After
- multiple logs
- failure to save to SA

You should break these out into separate test cases, with a shared setup function for any necessary context.